REGRESSION(283049@main): [ macOS EWS ]: imported/w3c/web-platform-tests/media-source/mediasource-addsourcebuffer-mode.html is a flaky crash

jyavenard opened the pull request:
#### c482824aa7fdab2efa87f41a8e811fe22d42c144
<pre>
REGRESSION(283049@main): [ macOS EWS ]: imported/w3c/web-platform-tests/media-source/mediasource-addsourcebuffer-mode.html is a flaky crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=279232">https://bugs.webkit.org/show_bug.cgi?id=279232</a>
<a href="https://rdar.apple.com/135376752">rdar://135376752</a>

Reviewed by Youenn Fablet.

In 283049@main we assumed that if we only had a single audio renderer, we could ignore the trackID as it is susceptible to change
over time.
However, a few places in the code assumes that if we can match an audio renderer to a trackID it indicates that we have a track configured.
However, the method could be called with a video&apos;s trackID and so we would incorrectly match the first audio renderer instead of returning nil.
This would cause later assertion as we would try to enqueue samples from an unexpected type.

We revert the change 283049@main ignoring the TrackID if we only had one audio track, going back to the previous logic.

We extend the scope of 283049@main to handling having the TrackID of the video tracks changing after a new init segment.

Covered by existing tests.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::updateTrackIds):
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeEnabled):

Canonical link: <a href="https://commits.webkit.org/283266@main">https://commits.webkit.org/283266@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bab3bdf18a010dd6a506f8ca62579429306b41fe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65733 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45106 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69759 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16342 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67851 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52905 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16624 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52763 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11343 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68800 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41641 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/56892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33392 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38316 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/14272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15218 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60163 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/14614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71465 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9688 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14044 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60081 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9720 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/56958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/60360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14489 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7995 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/1649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40914 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41990 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43173 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41734 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -934,10 +931,6 @@ static inline bool shouldAddContentKeyRecipients()
-        return m_audioRenderers.begin()->second.get();
     if (auto it = m_audioRenderers.find(trackID); it != m_audioRenderers.end())
         return it->second.get();
     return nil;
```
I would tend to write it as:
```
auto it = m_audioRenderers.find(trackID);
return it != m_audioRenderers.end() : it->second.get() : nil;
```
Also why not returning a RetainPtr?

jyavenard replied:
done

jyavenard committed:
REGRESSION(283049@main): [ macOS EWS ]: imported/w3c/web-platform-tests/media-source/mediasource-addsourcebuffer-mode.html is a flaky crash
https://bugs.webkit.org/show_bug.cgi?id=279232
rdar://135376752

Reviewed by Youenn Fablet.

In 283049@main we assumed that if we only had a single audio renderer, we could ignore the trackID as it is susceptible to change
over time.
However, a few places in the code assumes that if we can match an audio renderer to a trackID it indicates that we have a track configured.
However, the method could be called with a video's trackID and so we would incorrectly match the first audio renderer instead of returning nil.
This would cause later assertion as we would try to enqueue samples from an unexpected type.

We revert the change 283049@main ignoring the TrackID if we only had one audio track, going back to the previous logic.

We extend the scope of 283049@main to handling having the TrackID of the video tracks changing after a new init segment.

Covered by existing tests.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::updateTrackIds):
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeEnabled):

Canonical link: https://commits.webkit.org/283266@main
